pulsar-broker/src/main/java/org/apache/pulsar/broker/service/DrainingHashesTracker.java

```diff
@@ -0,0 +1,250 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.service;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * A thread-safe map to store draining hashes in the consumer.
+ */
+@Slf4j
+public class DrainingHashesTracker {
+    private final String dispatcherName;
+    private final UnblockingHandler unblockingHandler;
+    // optimize the memory consumption of the map by using primitive int keys
+    private final Int2ObjectOpenHashMap<DrainingHashEntry> drainingHashes = new Int2ObjectOpenHashMap<>();
+    int batchLevel;
+    boolean unblockedWhileBatching;
+
+    /**
+     * Represents an entry in the draining hashes tracker.
+     */
+    public static class DrainingHashEntry {
+        private final Consumer consumer;
+        private int refCount;
+        private int blockedCount;
+
+        /**
+         * Constructs a new DrainingHashEntry with the specified Consumer.
+         *
+         * @param consumer the Consumer instance
+         */
+        DrainingHashEntry(Consumer consumer) {
+            this.consumer = consumer;
+        }
+
+        /**
+         * Gets the consumer that contained the hash in pending acks at the time of creating this
+         * entry. Since a particular hash can be assigned to only one consumer at a time, this consumer
+         * cannot change. No new pending acks can be added in the {@link PendingAcksMap} when there's
+         * a draining hash entry for a hash in {@link DrainingHashesTracker}.
+         *
+         * @return the consumer instance that contained the hash in pending acks at the time of creating this entry
+         */
+        public Consumer getConsumer() {
+            return consumer;
+        }
+
+        /**
+         * Increments the reference count.
+         */
+        void incrementRefCount() {
+            refCount++;
+        }
+
+        /**
+         * Decrements the reference count.
+         *
+         * @return true if the reference count is zero, false otherwise
+         */
+        boolean decrementRefCount() {
+            return --refCount == 0;
+        }
+
+        /**
+         * Increments the blocked count.
+         */
+        void incrementBlockedCount() {
+            blockedCount++;
+        }
+
+        /**
+         * Checks if the entry is blocking.
+         *
+         * @return true if the blocked count is greater than zero, false otherwise
+         */
+        boolean isBlocking() {
+            return blockedCount > 0;
+        }
+
+        /**
+         * Returns a string representation of the DrainingHashEntry.
+         *
+         * @return a string representation of the DrainingHashEntry
+         */
+        public String toString() {
```

Comment:
Could be using `@ToString` lombok annotation

Reply:
Agree. Changed.